iree/compiler/Dialect/Stream/Transforms/ElideAsyncCopies.cpp

```diff
@@ -0,0 +1,531 @@
+// discard as they go out of scope. For now this strictly checks last use.
+class LastUseAnalysis {
+ public:
+  explicit LastUseAnalysis(Operation *rootOp)
+      : explorer(rootOp, TraversalAction::SHALLOW),
+        solver(explorer, allocator) {
+    explorer.setOpAction<IREE::Util::InitializerOp>(TraversalAction::RECURSE);
+    explorer.setOpAction<mlir::FuncOp>(TraversalAction::RECURSE);
+    explorer.setDialectAction<IREE::Stream::StreamDialect>(
+        TraversalAction::RECURSE);
+    // Ignore the contents of executables (linalg goo, etc).
+    explorer.setOpAction<IREE::Stream::ExecutableOp>(TraversalAction::IGNORE);
+    explorer.initialize();
+
+    assert(rootOp->getNumRegions() == 1 && "expected module-like root op");
+    topLevelOps = llvm::to_vector<4>(
+        rootOp->getRegions().front().getOps<mlir::CallableOpInterface>());
+  }
+
+  // Runs analysis and populates the state cache.
+  // May fail if analysis cannot be completed due to unsupported or unknown IR.
+  LogicalResult run() {
+    // Seed all block arguments throughout the program.
+    for (auto callableOp : getTopLevelOps()) {
+      for (auto &block : *callableOp.getCallableRegion()) {
+        for (auto arg : block.getArguments()) {
+          if (arg.getType().isa<IREE::Stream::ResourceType>()) {
+            solver.getOrCreateElementFor<ArgumentSemantics>(
+                Position::forValue(arg));
+          }
+        }
+      }
+    }
+
+    // Run solver to completion.
+    return solver.run();
+  }
+
+  // Returns a list of all top-level callable ops in the root op.
+  ArrayRef<mlir::CallableOpInterface> getTopLevelOps() const {
+    return topLevelOps;
+  }
+
+  // Returns true if block argument |arg| is passed in by-value/move (it's the
+  // last use from all callers/predecessor branches). When false the value
+  // represented by the argument may have other uses outside of its block.
+  bool isArgMoved(BlockArgument arg) {
+    auto argumentSemantics =
+        solver.lookupElementFor<ArgumentSemantics>(Position::forValue(arg));
+    if (!argumentSemantics) return false;
+    return argumentSemantics->getAssumedByValue();
+  }
+
+  // Returns true if |userOp| is the last user of |operand|.
+  bool isLastUser(Value operand, Operation *userOp) {
+    auto lastUsers =
+        solver.getOrCreateElementFor<LastUsers>(Position::forValue(operand));
+    return lastUsers.isAssumedLastUser(userOp);
+  }
+
+ private:
+  Explorer explorer;
+  llvm::BumpPtrAllocator allocator;
+  DFX::Solver solver;
+  SmallVector<mlir::CallableOpInterface> topLevelOps;
+};
+
+// Returns true if the given |operand| value does not need a copy on write.
+// This is a conservative check and will return false ("not safe to elide") in
+// many cases that otherwise don't need a copy. The
+// -iree-stream-elide-async-copies pass will do a whole-program analysis and
+// remove the copies we insert here when possible.
```

Comment:
This comment matches the one in the other file (for `isSafeToElideCOW`) pretty closely, and even references the pass that this file defines itself. Is that intentional?

Reply:
yeah they are both doing the same kind of check, though from different sides - one is testing if it can avoid inserting a copy while the other is testing if it can remove one that exists - in the former we don't have full information (as we are iteratively inserting new clones and at any particular site the subsequent uses aren't available yet for analysis) while in the latter we have the whole module available and are using the solver to iterate over all state

(comment is similar because both started as the same pass and I forked them out :)